### ISSUE TITLE
Rust implementation of systemd-stub

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#26640]
As we have planned for some time (old PoC: https://github.com/systemd/systemd/pull/19598 ), we wish to introduce Rust into the codebase. We have a volunteer to start from systemd-stub, which is nicely self-contained and isolated.

Requirements:
- track unstable status of uefi-rs crate: https://github.com/rust-osdev/uefi-rs/issues/452 close to becoming stable
- use meson as the build system
- (might require https://github.com/mesonbuild/meson/pull/11213 and https://github.com/rust-lang/rust/pull/106560 )
- dependencies need to be sensible and meet the same bar as the C ones, ie: they are maintained, reliable, do not break API
- dependencies need to be available in distros, if not, list them and file requests (eg: uefi-rs) on the main distros
- all supported architectures need to work (x86/x86_64/armv7/armv8/riscv64)
- integrated in the code base, ie: share common internal utility functions both ways as it is done now with the C implementation

Action plan:
- work will start out of tree to get started quickly, but once basics are working, it will be integrated in the main repository
- both implementations will be available in the repository at the same time for at least a few releases
- both implementations will be buildable at the same time, with a new option for the rust one for at least a few releases
- testing: limited infrastructure for sd-stub, mostly through the mkosi CI jobs

**Comment by YHNdnzj** [comment]
FWIW, here's one WIP implementation from NixOS community: https://github.com/nikstur/sd-stub-rs

**Comment by bluca** [comment]
Yes, @nikstur is the developer who volunteered to do the implementation

**Comment by nikstur** [comment]
cc @blitz

**Comment by medhefgo** [comment]
rust 😻

Since I dunno if this was brought up:
 - You should be aware that rust only has aarch64, x86 and x86_64 because it uses the `*-windows-gnu` llvm targets. Also, gccrs is not gonna improve that. For other arches one would either have to teach llvm a coff target triple or build to an ELF target and convert that to PE.
- Similarly when sharing c and rust code: You have to be aware that our c code is built to ELF/sysv ABI, while the UEFI rust side would target a windows ABI. You cannot link those together. So you need to choose one ABI to target (I believe build.rs can compile the c code for us so it has the right one?)

**Comment by phip1611** [comment]
> Similarly when sharing c and rust code: You have to be aware that our c code is built to ELF/sysv ABI, while the UEFI rust side would target a windows ABI. You cannot link those together. So you need to choose one ABI to target (I believe build.rs can compile the c code for us so it has the right one?)

When you create the bindings for FFI code, you can specify the ABI in Rust. My GitHub also provides an [example for that](https://github.com/phip1611/rust-different-calling-conventions-example/blob/main/src/main.rs).

So you can use 

```rust
extern "sysv64" {
    fn foobar(a: i64, b: i64) -> i64;
}
```

and link foobar from a library produced by a C compiler.

**Comment by medhefgo** [comment]
I am well aware that rust lets you select a calling convention, its ffi would be borderline useless without it. I am talking about the *actual* ABI (it's more than just a calling convention). You *cannot* link stuff built for a pe/coff target with stuff built for elf/sysv. Even *if* the linker lets you, you're inviting things to break in weird ways.

**Comment by phip1611** [comment]
> I am well aware that rust lets you select a calling convention, its ffi would be borderline useless without it. I am talking about the _actual_ ABI (it's more than just a calling convention). You _cannot_ link stuff built for a pe/coff target with stuff built for elf/sysv. Even _if_ the linker lets you, you're inviting things to break in weird ways.

Interesting. I never did this, but I thought it is possible.

> you're inviting things to break in weird ways.

Is this something that is worth prototyping and testing? Do we have a guarantee that one break things in weird ways with that approach or is this just an assumption/expectation that can be refuted with good test coverage against the compiled binary?

**Comment by medhefgo** [comment]
> Interesting. I never did this, but I thought it is possible.

Easy to trigger and I doubt you'll get any sympathy for that crash:
```sh
$ cat test1.c
void a(void) {
        __builtin_printf("a\n");
}

$ cat test2.c
void b(void) {
        __builtin_printf("b\n");
}

$ cat test3.c
void a(void);
void b(void);
int main(int argc, char *argv[]) {
        a();
        b();
        return 0;
}

$ gcc -c test1.c -o test1.o
$ x86_64-w64-mingw32-gcc -c test2.c -o test2.o
$ gcc -c test3.c -o test3.o
$ gcc -o test test3.o test2.obj test1.o 
/usr/bin/ld: BFD (GNU Binutils) 2.40 assertion fail /usr/src/debug/binutils/binutils-gdb/bfd/reloc.c:8632
test2.o:test2.c:(.pdata+0x0): dangerous relocation: collect2: fatal error: ld terminated with signal 11 [Segmentation fault], core dumped
compilation terminated.
$ clang -fuse-ld=lld -o test test3.o test2.o test1.o 
ld.lld: error: test2.o: unknown file type
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

> Is this something that is worth prototyping and testing? Do we have a guarantee that one break things in weird ways with that approach or is this just an assumption/expectation that can be refuted with good test coverage against the compiled binary?

This is not worth anyone's time. Just compile both sides for the same target triple and avoid the whole thing.

**Comment by soloturn** [comment]
@danielrh did make https://github.com/dropbox/rust-brotli so that it can be dropped in for c written brotli. you mean doing something like he, @medhefgo ?

**Comment by medhefgo** [comment]
I'm saying that rust/cargo either has to build c stuff on our behalf for the uefi target (or give us the necessary cc args) or rust code has to be build for an ELF target and produce a static PIC library. Also, when building for ELF on x86 care has to be taken to disable mmx/sse etc, because firmwares are stupid.

**Comment by safinaskar** [comment]
Please, chose build everything as PE (as opposed to building everything as ELF). I have two arguments:

- Building everything as ELF is wrong for these reasons: https://dvdhrm.github.io/2019/01/31/goodbye-gnuefi
- Upstream Rust compiler chose PE, not ELF. The only EFI-related x86_64 target I was able to find on https://rust-lang.github.io/rustup-components-history/ is x86_64-unknown-uefi, and it is based on PE as documented here: https://doc.rust-lang.org/nightly/rustc/platform-support/unknown-uefi.html

**Comment by RaitoBezarius** [comment]
> You're also failing to realize that other arches than the three (and a half) PE arches supported properly by llvm exist. As long as there is no proper and easy to use toolchain and build systems for all the EFI arches, using the conversion process is far superior. Especially when it comes to mixing C and rust object files.

Interesting, I have been working into trying properly systemd-(stub|boot) and our NixOS UEFI stuff on RISC-V but AFAIK, it's impossible to have proper LLVM support there because Windows has not specified how COFF relocations should work on RISC-V.

I know that we can still compile to ELF then convert ELF -> PE relocations because those are cross platforms and generic, but it's a bit disturbing to me that any attempt to have more complicated programs have to go through the ELF process because we are blocked on MSFT specifying those. Wouldn't it be better to have the conversion process built-in in a compiler, e.g. LLVM while explaining our case? I would imagine it's somewhat of a linker job to perform this kind of thing.

---

Also, is there a place that says all the "EFI" arches somewhere? AFAIK: i686, x86_64, arm, aarch64, riscv64 are the obvious one atm. All of them except riscv64 are supported I believe.

I know about some exotic PPC64 stuff but are not officially supported by the specification right? Are those in scope?

**Comment by bluca** [comment]
I seem to remember that loongarch is also going to target EFI

**Comment by medhefgo** [comment]
Technically, llvm supports arm for windows, but it sets a machine ID in the PE header that is refused by EDKII. Such a binary can be modified post-build and it seems to boot fine in qemu, but someone with actual knowledge of these things should update llvm to make it possible. It's presumably why there is no rust uefi arm target.

Right now we fully support ia32/x64/arm/aa64 EFI arches. There is also compile-tested riscv/loongarch support.

> Wouldn't it be better to have the conversion process built-in in a compiler, e.g. LLVM while explaining our case? I would imagine it's somewhat of a linker job to perform this kind of thing.

Conversion in the toolchain is ugly and not what you want when said toolchain is in the perfect position to fix this properly. What you *really* want is for *toolchains* to have actual UEFI target triplets for proper cross-building similar to rust/zig UEFI targets. The difference would be that they'd be defined on their own and not piggy-back on windows targets.

I also don't see how COFF relocations matter here. I'd consider the object format for compilation artifacts as an implementation detail. llvm could use whatever format or extensions to it and simply declare the files as opaque to other consumers. For the final PE image the riscv/loongarch base relocations are already well defined in the PE spec.

**Comment by RaitoBezarius** [comment]
> Technically, llvm supports arm for windows, but it sets a machine ID in the PE header that is refused by EDKII. Such a binary can be modified post-build and it seems to boot fine in qemu, but someone with actual knowledge of these things should update llvm to make it possible. It's presumably why there is no rust uefi arm target.
> 
> Right now we fully support ia32/x64/arm/aa64 EFI arches. There is also compile-tested riscv/loongarch support.

Got it.

> 
> > Wouldn't it be better to have the conversion process built-in in a compiler, e.g. LLVM while explaining our case? I would imagine it's somewhat of a linker job to perform this kind of thing.
> 
> Conversion in the toolchain is ugly and not what you want when said toolchain is in the perfect position to fix this properly. What you _really_ want is for _toolchains_ to have actual UEFI target triplets for proper cross-building similar to rust/zig UEFI targets. The difference would be that they'd be defined on their own and not piggy-back on windows targets.

I don't know about that "perfect position". We had some serious attempts to untangle things in https://github.com/NixOS/nixpkgs/pull/228374 and the whole dance about GNU vs LINK.exe style flags made me think that it would be hard to have proper targets on their own.

But ultimately, I agree I wished we had "obvious targets" for those. It's unclear how they fit in the GNU or LLVM triples model and what is the kernel of UEFI or what is the ABI of UEFI, etc.

> 
> I also don't see how COFF relocations matter here. I'd consider the object format for compilation artifacts as an implementation detail. llvm could use whatever format or extensions to it and simply declare the files as opaque to other consumers. For the final PE image the riscv/loongarch base relocations are already well defined in the PE spec.

See https://github.com/rust-osdev/uefi-rs/issues/241 / https://reviews.llvm.org/D145539 for context on that. I am very interested if you have compelling arguments to un-deadlock the situation. I am willing to spend time to fix the patches upstream and have the proper solution.

**Comment by medhefgo** [comment]
> I don't know about that "perfect position". We had some serious attempts to untangle things in [NixOS/nixpkgs#228374](https://github.com/NixOS/nixpkgs/pull/228374) and the whole dance about GNU vs LINK.exe style flags made me think that it would be hard to have proper targets on their own.

What does linker flag style have to do with this?

> But ultimately, I agree I wished we had "obvious targets" for those. It's unclear how they fit in the GNU or LLVM triples model and what is the kernel of UEFI or what is the ABI of UEFI, etc.

There is no kernel or libc, that's why the target has an `unknown` in there. And the ABI is as defined by the UEFI spec.

> See [rust-osdev/uefi-rs#241](https://github.com/rust-osdev/uefi-rs/issues/241) / https://reviews.llvm.org/D145539 for context on that. I am very interested if you have compelling arguments to un-deadlock the situation. I am willing to spend time to fix the patches upstream and have the proper solution.

Pick your poison:
 - get MS to define COFF relocations for riscv/loongarch
 - add a custom object file format
 - teach linkers how to link ELF object files into PE images
 - add/use private COFF relocation definitions (possibly as detached from the COFF object file)

**Comment by RaitoBezarius** [comment]
> > I don't know about that "perfect position". We had some serious attempts to untangle things in [NixOS/nixpkgs#228374](https://github.com/NixOS/nixpkgs/pull/228374) and the whole dance about GNU vs LINK.exe style flags made me think that it would be hard to have proper targets on their own.
> 
> What does linker flag style have to do with this?

"Proper targets" assume you choose a linker flag style, see what Rust is doing: https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/aarch64_unknown_uefi.rs#L11

Depending on your linker flag style, there are different effects in (meta) build systems because they may assume certain linker styles, in Nix, we had only GNU style flags, and we had purity detection based on the presence of foreign paths, which broke the handling for LINK.exe style flags because `/NOLOGO` was misdirected as a path.
Obviously, this is not an inherent limitation, but at least, there is no "the proper target" but multiple proper targets even on, let's say, x86_64-linux.

See https://github.com/rust-lang/rust/pull/110869 also.

> 
> > But ultimately, I agree I wished we had "obvious targets" for those. It's unclear how they fit in the GNU or LLVM triples model and what is the kernel of UEFI or what is the ABI of UEFI, etc.
> 
> There is no kernel or libc, that's why the target has an `unknown` in there. And the ABI is as defined by the UEFI spec.

Oh sorry, I agree there is no kernel or libc. But do we agree that https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/aarch64_unknown_uefi.rs#L15 indicate that the target has unknown kernel and `windows` as a vendor? How do you parse `windows` here?

Does it have a name or is it called `uefi` ABI mapped to whatever makes sense on the target platform?

> 
> > See [rust-osdev/uefi-rs#241](https://github.com/rust-osdev/uefi-rs/issues/241) / https://reviews.llvm.org/D145539 for context on that. I am very interested if you have compelling arguments to un-deadlock the situation. I am willing to spend time to fix the patches upstream and have the proper solution.
> 
> Pick your poison:
> 
> * get MS to define COFF relocations for riscv/loongarch
> * add a custom object file format
> * teach linkers how to link ELF object files into PE images

Why would this one be a bad idea in your opinion?

> * add/use private COFF relocation definitions (possibly as detached from the COFF object file)

Interesting idea!

**Comment by medhefgo** [comment]
> Oh sorry, I agree there is no kernel or libc. But do we agree that https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/aarch64_unknown_uefi.rs#L15 indicate that the target has unknown kernel and `windows` as a vendor? How do you parse `windows` here?

The idea would be for llvm to provide a UEFI target triplet so that rust and others wouldn't have to and to eliminate any confusion.

> Does it have a name or is it called `uefi` ABI mapped to whatever makes sense on the target platform?

I would follow the rust style for a gcc/llvm triplet: `$ARCH-unknown-uefi`. I think technically, the unknown is the platform and uefi the ABI.

> Why would this one be a bad idea in your opinion?

It's a bad idea if the input objects have different ABI (one built for UEFI/windows, one for ELF/sysv). But if all ELF files involved were built for the same ABI (ideally the new UEFI target triplet) it should not pose a problem per-se. The issue, I guess, is that people (me included) tend to conflate object/image formats and the ABI of the *code* within. They really are two independent things.

In fact, you can already sort of do this already with bfd for x86 (it's called linker emulation):
```
$ cat a.c
void a(void){}
$ cat b.c
void b(void){}
$ gcc -c a.c
$ gcc -c b.c
$ ld a.o b.o -mi386pep -o c
$ file a.o b.o c
a.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
b.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
c:   PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows, 5 sections
```

The issue with this is that it's poorly integrated everywhere. gcc likes to pass the wrong link args because it thinks it's going to link an ELF file, so does meson. And I imagine other parts in the toolchain have the same issue, hence why I prefer the actual UEFI target triplet to sidestep these things. It's also x86 only (the new aarch64 emulation afaik doesn't do relocations, so it's for objcopy only). And I imagine it's also poorly tested in general because of it.

**Comment by RossComputerGuy** [comment]
> I would follow the rust style for a gcc/llvm triplet: `$ARCH-unknown-uefi`. I think technically, the unknown is the platform and uefi the ABI.

Not quite, this looks like an `$arch-$vendor-$os` triplet. UEFI would be more akin to an OS and not an ABI here.

Also, I have other news to share which is relevant here:
1. LLVM 20 is starting to support UEFI OS targets more, we're landing some fixes which gets clang to compile things
2. LLVM libc is starting to get support for UEFI, this means a POSIX wrapper around UEFI protocols and stuff

I know the libc one might not be relevant to Rust but does at least mean systemd might be able to clean up the code if LLVM libc could be integrated in.

**Comment by DaanDeMeyer** [comment]
> > I would follow the rust style for a gcc/llvm triplet: `$ARCH-unknown-uefi`. I think technically, the unknown is the platform and uefi the ABI.
> 
> Not quite, this looks like an `$arch-$vendor-$os` triplet. UEFI would be more akin to an OS and not an ABI here.
> 
> Also, I have other news to share which is relevant here:
> 1. LLVM 20 is starting to support UEFI OS targets more, we're landing some fixes which gets clang to compile things
> 2. LLVM libc is starting to get support for UEFI, this means a POSIX wrapper around UEFI protocols and stuff
> 
> I know the libc one might not be relevant to Rust but does at least mean systemd might be able to clean up the code if LLVM libc could be integrated in.

Our current approach is compiler independent which we'd lose by starting to use this as well as depending on an extremely new LLVM so I don't think we'll be able to adopt this any time soon.
